### ISSUE TITLE
Add runtime check for `c_p_pointer` behaviour? 

### ISSUE THREAD

**Issue by Smit-create** [lcompilers/lpython#676]
I was testing pointers in `lpython` and used the following python code:
```py
from ltypes import c_p_pointer, CPtr, pointer, i16, Pointer

def f():
    yq: CPtr
    yptr1: Pointer[i16[:]]
    y: i16[2]
    y[0] = 1
    y[1] = 2
    yptr1 = pointer(y)
    # This works
    print(yptr1[0], yptr1[1])
    assert yptr1[0] == 1
    assert yptr1[1] == 2
    c_p_pointer(yq, yptr1)

    # Shouldn't this work the same way as earlier?
    print(yptr1[0], yptr1[1])
f()
```

Running this I get:
```
% lpython expr.py && ./a.out
1 2
zsh: segmentation fault  ./a.out
```

**Comment by certik** [comment]
Can you specify what you currently get, and what is the expected behavior?

**Comment by Smit-create** [comment]
> Can you specify what you currently get, and what is the expected behavior?

Yeah, updated the OP

**Comment by certik** [comment]
My understanding is that `c_p_pointer` assigns `yq` to `yptr1`. But `yq` points to random memory, so I would expect it to segfault, which it does.

**Comment by czgdp1807** [comment]
> ```python
> # Shouldn't this work the same way as earlier?
>     print(yptr1[0], yptr1[1])
> ```

No it shouldn't. `yptr1`'s data element now points to the same location where `yq` points. `yptr1` is no longer linked to `y`. So segmentation fault or any random behaviour is completely normal here.

**Comment by Smit-create** [comment]
Understood. Thanks, @certik and @czgdp1807.

**Comment by certik** [comment]
Yes, it's an explanation why it segfaults. However, I would like to figure out how not to segfault in Debug mode. To do that, we should add runtime checks for dangling pointers (`yq` above), and if used (in `print(yptr1[0], yptr1[1])` above), it should give a runtime stacktrace, saying that the pointer is dangling, so it cannot be used.

This is not urgent, and I think we might want to tackle this a little bit later, but it's something I definitely want to do.

**Comment by czgdp1807** [comment]
I see. I will look into it though at the bottom of my priority list right now. Let me know if it should be dealt with on an urgent basis. Thanks.

**Comment by certik** [comment]
This is not urgent at all. I am thinking perhaps in 6 months we can start adding these runtime checks (for bounds checking, stack overflow, dangling pointers, etc.). Right now let's get LPython into alpha and LFortran into beta.
